src/shogun/modelselection/NewGridSearch.h

```diff
@@ -0,0 +1,354 @@
+		    const std::string& param,
+		    const std::shared_ptr<ParameterNode>& node);
+
+		/**
+		 * Attach a new node at a given location in the tree
+		 * defined by the parameter name
+		 * @param name
+		 * @param node
+		 */
+		ParameterNode* attach(const std::string& param, ParameterNode* node);
+
+		/**
+		 * Attach a new value at a given location in the tree
+		 * defined by the parameter name
+		 * @param name
+		 * @param node
+		 */
+		template <typename T>
+		ParameterNode* attach(const std::string& param, T value)
+		{
+			if (!set_param_helper(param, make_any(value)))
+				SG_SERROR("Could not attach %s", param.c_str());
+			return this;
+		}
+
+		/**
+		 * Returns the next parameter combination.
+		 *
+		 * @return
+		 */
+		ParameterNode* next_combination()
+		{
+			return get_next();
+		}
+
+		virtual const char* get_name() const noexcept
+		{
+			return "ParameterNode";
+		}
+
+		/**
+		 * Returns the name of the node which is
+		 * infered by the CSGObject
+		 * @return
+		 */
+		const char* get_parent_name() const
+		{
+			return m_parent->get_name();
+		}
+
+		std::string to_string() const noexcept;
+
+		/**
+		 * Replaces a node in a specific position of the tree node.
+		 * A tree node can contain several nodes which represent
+		 * different objects that can be used by the parent object.
+		 *
+		 * @param node_name
+		 * @param index
+		 * @param node
+		 */
+		void replace_node(
+		    const std::string& node_name, size_t index,
+		    const std::shared_ptr<ParameterNode>& node);
+
+		size_t n_nodes()
+		{
+			return m_nodes.size();
+		}
+
+	protected:
+		/**
+		 * Creates a node that mimics the CSGObject structure
+		 * and places it in the tree using the given name.
+		 * This is meant to be only used inside constructors,
+		 * as it doesn't check if the node should exist.
+		 * @param name
+		 * @param obj
+		 */
+		virtual void create_node(const std::string& name, CSGObject* obj);
+		/**
+		 * Internal method to set the value (with Any type) of a node
+		 *
+		 * @param param
+		 * @param value
+		 * @return
+		 */
+		virtual bool
+		set_param_helper(const std::string& param, const Any& value);
+		/**
+		 * Internal method to set the node within a node
+		 *
+		 * @param param
+		 * @param value
+		 * @return
+		 */
+		bool set_param_helper(
+		    const std::string& param,
+		    const std::shared_ptr<ParameterNode>& value);
+		/**
+		 * Internal method to get the next combination. In the base class
+		 * this is the current set, as it is the only parameter combination
+		 * available.
+		 *
+		 * @return
+		 */
+		virtual ParameterNode* get_next()
+		{
+			return get_current();
+		}
+		/**
+		 * Internal method to get the current parameter combination
+		 * @return
+		 */
+		virtual ParameterNode* get_current();
+
+		/** vector of child nodes */
+		std::map<std::string, std::vector<std::shared_ptr<ParameterNode>>>
+		    m_nodes;
+		/** pointer to object that is being mimicked */
+		CSGObject* m_parent;
```

Comment:
@vigsterkr do you think i should use a raw pointer? This is pointing to a model that is passed to the constructor (where i increment the counter). Would it make sense to use `std::unique_ptr`? Right now I unreference this pointer in the destructor.